### ISSUE TITLE
Handle Let's Encrypt's lack of CRLs more gracefully

### ISSUE THREAD

**Issue by mpalmer** [crtsh/certwatch_db#71]
Finding unrevoked certificates for most CAs is straightforward -- join on crl_revoked (or use `certificate_lifecycle`) and you're off to the races.  However, Let's Encrypt doesn't publish CRLs, and so we're left calling `ocsp_embedded` over and over and over and over...

Since revocation is "one-way", in that once your OCSP responder says "yep this one's busted" you can't take it back, it would be handy (I think) if `ocsp_embedded` calls that returned a revoked status cached that info in a table (perhaps `revoked_by_ocsp`?) that you could join on.  I wouldn't recommend `ocsp_embedded` using that table itself for lookups, because that would change its semantics, but having a table of "these are certs we know OCSP has said are revoked" would allow my code to be far more efficient (as in, hammer the crt.sh database a lot less), because at the moment it does something like this:

```
for each unrevoked, unexpired cert
    if it's an LE cert
        hit crt.sh and do an ocsp_embedded() lookup for that cert
    fi
done
```

At the moment I've revoked a few hundred LE certs for having compromised private keys; I don't expect that number to go down.  Not having to do several hundred unnecessary `ocsp_embedded` calls every time I drag the crt.sh database can only be good for everyone.

This also, I *think*, is causing problems for misissued.com; the batches I've just created (https://misissued.com/batch/64 and https://misissued.com/batch/65) don't show any of the LE certs as "Revoked", however if you load them up in crt.sh and ask for an OCSP lookup, they're all revoked.

If all this sounds vaguely sensible, but you're lacking time to implement it, I can have a crack at a PR, however I must warn you in advance that my SQL skills are not `certwatch_db` grade...

**Comment by robstradling** [comment]
> Since revocation is "one-way", in that once your OCSP responder says "yep this one's busted" you can't take it back

That's true for revocation via CRL, because the BRs (4.9.13) prohibit the use of certificate suspension.

However, sadly I can find no definitive rule that would prevent an OCSP-only CA from unrevoking certificates.

**Comment by robstradling** [comment]
Also, the slave instances of the crt.sh database (which is what your `ocsp_embedded` calls are serviced by) are read-only, and there isn't currently any mechanism for the slaves to feed back information (e.g., regarding revocation checks) to the master.

**Comment by robstradling** [comment]
BTW, have you considered pulling the cert and CA cert details from crt.sh, but then performing the OCSP checks (and caching the results) from your end?

**Comment by mpalmer** [comment]
> However, sadly I can find no definitive rule that would prevent an OCSP-only CA from unrevoking certificates.

:keanu: Whoa.  That's a bit of a mind-bender.  Although I was also surprised to discover that CRLs aren't even mandatory.  It's been a big week.

> Also, the slave instances of the crt.sh database (which is what your ocsp_embedded calls are serviced by) are read-only, and there isn't currently any mechanism for the slaves to feed back information (e.g., regarding revocation checks) to the master.

That would, indeed, put something of a crimp in that particular plan.

> BTW, have you considered pulling the cert and CA cert details from crt.sh, but then performing the OCSP checks (and caching the results) from your end?

I have, but that would involve writing code of my own, and I am an extremely lazy person.  :grinning:

**Comment by RufusJWB** [comment]
This will be an interesting Ballot :-)
